refactor(ast, regular_expression): shorten `ContentEq` implementations

overlookmotel opened the pull request:
Shorten codegen-ed implementations of `ContentEq` for enums, matching what Rust produces for `#[derive(PartialEq)]` (inspired by #8517).

overlookmotel commented:
* **#8519** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8519?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8519?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=8519).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented:
@camc314 I half-hoped this would produce a speed-up too, but no. I think `ContentEq` is just inherently expensive, especially when the 2 AST fragments being compared are large and *mostly* the same, so you have to crawl a long way through the structures to find out if they're the same or not.

Boshen commented:
No change in benchmark, so merging.

Boshen commented:
### Merge activity

* **Jan 17, 12:58 AM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.
* **Jan 17, 12:58 AM EST**: A user added this pull request to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc).

overlookmotel committed:
refactor(ast, regular_expression): shorten `ContentEq` implementations (#8519)

Shorten codegen-ed implementations of `ContentEq` for enums, matching what Rust produces for `#[derive(PartialEq)]` (inspired by #8517).

overlookmotel commented:
> it might be worth double checking this. but i think this approach has worse performance than before. If we consider "left" and "other", if "left" matches the first enum variant on the first thing of the match statement, and right does not, it will try ever match pattern rather than bailing out early. << this might not be correct, but it's what i understood from godbolt and some rust forums

I would assume compiler is smart enough to see that each match arm is mutually exclusive, and so will turn it into "exit if discriminants are not the same, then jump table" rather than looping through each condition. I would have thought we can safely assume that what Rust produces for `PartialEq` is optimal. If it wasn't, it would have been corrected long ago, as it's such a common construct and critical for performance.

That's all assumption though. I'd be interested to see what you saw in Godbolt which suggested otherwise.
